src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -113,42 +115,40 @@ ExpressionUnorderedSet ColumnPruningRule::_collect_actually_used_columns(const s
                                                    const ExpressionUnorderedSet& referenced_columns) {
+  std::vector<std::shared_ptr<AbstractLQPNode>> leafs;
+
   // Collect all parents of leaves and on which input side their leave is
   // (if a node has two leaves as inputs, it will be collected twice)
   auto leaf_parents = std::vector<std::pair<std::shared_ptr<AbstractLQPNode>, LQPInputSide>>{};
   visit_lqp(lqp, [&](auto& node) {
-    for (const auto input_side : {LQPInputSide::Left, LQPInputSide::Right}) {
-      const auto input = node->input(input_side);
-      if (input && input->input_count() == 0) leaf_parents.emplace_back(node, input_side);
+    // Early out for non-leafs
+    if (node->input_count() != 0) {
+      return LQPVisitation::VisitInputs;
     }
 
-    // Do not visit the ProjectionNode we may have just inserted, that would lead to infinite recursion
-    return LQPVisitation::VisitInputs;
-  });
-
-  // Insert ProjectionNodes that prune unused columns between the leaves and their parents
-  for (const auto& parent_and_leaf_input_side : leaf_parents) {
-    const auto& parent = parent_and_leaf_input_side.first;
-    const auto& leaf_input_side = parent_and_leaf_input_side.second;
-    const auto leaf = parent->input(leaf_input_side);
-
-    // Collect all columns from the leaf that are actually referenced
-    auto referenced_leaf_columns = std::vector<std::shared_ptr<AbstractExpression>>{};
-    for (const auto& expression : leaf->column_expressions()) {
-      if (referenced_columns.find(expression) != referenced_columns.end()) {
-        referenced_leaf_columns.emplace_back(expression);
+    auto pruned_column_ids = std::vector<ColumnID>{};
+    for (const auto& expression : node->column_expressions()) {
+      if (referenced_columns.find(expression) == referenced_columns.end()) {
+        const auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
+        pruned_column_ids.emplace_back(column_expression->column_reference.original_column_id());
       }
     }
 
-    if (leaf->column_expressions().size() == referenced_leaf_columns.size()) continue;
+    // We cannot create a Table without columns - since Chunks rely on their first column to determine their row count
+    if (pruned_column_ids.size() == node->column_expressions().size() && !pruned_column_ids.empty()) {
+      pruned_column_ids.pop_back();
+    }
 
-    // We cannot have a ProjectionNode that outputs no columns, so let's avoid that
-    if (referenced_leaf_columns.empty()) continue;
+    if (const auto stored_table_node = std::dynamic_pointer_cast<StoredTableNode>(node)) {
+      stored_table_node->set_pruned_column_ids(pruned_column_ids);
+    } else if (const auto mock_node = std::dynamic_pointer_cast<MockNode>(node)) {
+      mock_node->set_pruned_column_ids(pruned_column_ids);
+    } else {
+      // We don't know how to prune columns from this node type (CreateViewNode, etc.), so do nothing
```

Comment:
CreateViewNode seems to be a weird example to choose here

Reply:
StoredTable/Mock/Create* are the only valid leaf-nodes and the former two can be pruned from, so this makes sense, imo

Reply:
Changing it to "leaf-node type" was enough to clarify.